presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,295 @@
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.expressions.LogicalRowExpressions;
+import com.facebook.presto.expressions.RowExpressionRewriter;
+import com.facebook.presto.expressions.RowExpressionTreeRewriter;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.FunctionAndTypeManager;
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.InputReferenceExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.spi.relation.VariableReferenceExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.facebook.presto.sql.relational.RowExpressionDeterminismEvaluator;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeCaseExpressionPredicate;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.expressions.LogicalRowExpressions.and;
+import static com.facebook.presto.expressions.LogicalRowExpressions.or;
+import static com.facebook.presto.expressions.LogicalRowExpressions.replaceArguments;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.SWITCH;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.WHEN;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Re-writes CASE expression predicate into a series of OR clauses
+ * <p>
+ * (CASE
+ * WHEN expression=constant1 THEN result1
+ * WHEN expression=constant2 THEN result2
+ * ELSE elseResult
+ * END) = value
+ * <p>
+ * becomes
+ * <p>
+ * (result1 = value AND expression=constant1) OR
+ * (result2 = value AND expression=constant2) OR
+ * (elseResult = value AND !(expression=constant1) AND !(expression=constant2))
+ * <p>
+ * This conversion happens when all of the following criteria are met:
+ * 1. Value is either a constant or column reference or input reference and not any function
+ * 2. The LHS expression in all WHEN clauses are the same.
+ *    For example, if one WHEN clause has a expression using col1 and another using col2, it will not work
+ * 3. The relational operator in the WHEN clause is equals. With other operators it is harder to check for exclusivity.
+ * 4. All the RHS expressions are a constant and unique
```

Comment:
If we do the rewrite without this check (the RHS is not unique)
```
(case 
when col1=1 return 'a'
when col1=1 return 'b'
else 'c') = 'b'
```
will get simplified to
```
(col1=1 AND 'a'='b') OR
(col1=1 AND 'b'='b') OR
('c' = 'b' AND col1<>1 AND col2<>2)
```
`'a'='b'` condition would evaluate to false, and will get simplified to
`col1=1`
But this is a wrong simplification. If col1=1, it would have gone to first WHEN clause in CASE statement and returned 'a' which wouldn't be equal to 'b'. Our simplified condition allows rows with col1=1 to pass but in actual case it should not pass.